task_manager/task_manager.hh

```diff
@@ -0,0 +1,308 @@
+/*
+ * Copyright (C) 2022-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+#pragma once
+
+#include "replica/database.hh"
+
+using task_id = utils::tagged_uuid<struct task_id_tag>;
+using is_abortable = bool_class <struct abortable_tag>;
+
+class task_manager : public seastar::peering_sharded_service<task_manager> {
+public:
+    class task;
+    class module;
+    struct config {
+        utils::updateable_value<uint32_t> task_ttl = utils::updateable_value<uint32_t>(0);
+    };
+    using task_ptr = seastar::lw_shared_ptr<task_manager::task>;
+    using task_map = std::unordered_map<task_id, task_ptr>;
+    using foreign_task_ptr = seastar::foreign_ptr<task_ptr>;
+    using foreign_task_vector = std::vector<foreign_task_ptr>;
+    using module_ptr = seastar::lw_shared_ptr<module>;
+    using modules = std::unordered_map<std::string, module_ptr>;
+private:
+    task_map _all_tasks;
+    modules _modules;
+    config _cfg;
+    abort_source& _as;
+    serialized_action _update_task_ttl_action;
+    utils::observer<uint32_t> _task_ttl_observer;
+    uint32_t _task_ttl;
+public:
+    class task_not_found : public std::exception {
+        sstring _cause;
+    public:
+        explicit task_not_found(task_id tid)
+            : _cause(format("task with id {} not found", tid))
+        { }
+
+        virtual const char* what() const noexcept override { return _cause.c_str(); }
+    };
+
+    enum class task_state {
+        created,
+        running,
+        done,
+        failed
+    };
+
+    class task {
```

Comment:
: public enable_lw_shared_from_this

Reply:
consider registering the task in the constructor, as it has access to the module via the impl parameter.